sound/soc/sof/topology.c

```diff
@@ -2464,15 +2465,198 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
 	.bytes_ext_ops_count = ARRAY_SIZE(sof_dspless_bytes_ext_ops),
 };
 
+enum tplg_device_id {
+	TPLG_DEVICE_SDW_JACK,
+	TPLG_DEVICE_SDW_AMP,
+	TPLG_DEVICE_SDW_MIC,
+	TPLG_DEVICE_PCH_DMIC,
+	TPLG_DEVICE_HDMI,
+	TPLG_DEVICE_MAX
+};
+
+#define SDCA_DEVICE_MASK (BIT(TPLG_DEVICE_SDW_JACK) | BIT(TPLG_DEVICE_SDW_AMP) | \
+			  BIT(TPLG_DEVICE_SDW_MIC))
+
+struct topology_file {
+	char *device;
+	char *file;
+	int be_id;
+	int dev;
+};
+
+/* The code is from https://stackoverflow.com/questions/47116974/remove-a-substring-from-a-string-in-c */
+static char *strremove(char *str, const char *sub)
+{
+	size_t len = strlen(sub);
+
+	if (len > 0) {
+		char *p = str;
+		size_t size = 0;
+
+		while ((p = strstr(p, sub)) != NULL) {
+			size = (size == 0) ? (p - str) + strlen(p + len) + 1 : size - len;
+			memmove(p, p + len, size - (p - str));
+		}
+	}
+	return str;
+}
+
 int snd_sof_load_topology(struct snd_soc_component *scomp, const char *file)
 {
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+	struct topology_file tplg_files[TPLG_DEVICE_MAX];
```

Comment:
We can ever have 1 instance providing a function? no two jack/amp/mic/dmic/hdmi ? only ever one from each?

Reply:
think about karaoke devices with multiple jacks/mics for example.

Reply:
The idea is to have a topology for a function. Like we have 3 hdmi pcms, but we only have 1 hdmi topology.